File: ios/LibraryImageSize.m
//
//  LibraryImageSize.m
//  MediaLibrary
//

#import "LibraryImageSize.h"
#import "MediaAssetManager.h"
#import <React/RCTConvert.h>
#import <React/RCTImageSource.h>
#import <React/RCTImageUtils.h>

NSURL *toFilePath(NSString *path) {
    if ([path hasPrefix:@"file://"]) {
        return [NSURL URLWithString:path];
    }
    return [NSURL URLWithString:[NSString stringWithFormat:@"file://%@", path]];
}

void ensurePath(NSString *path) {
    NSString *folderPath = [path stringByDeletingLastPathComponent];
    NSFileManager *fm = [NSFileManager defaultManager];
    [fm createDirectoryAtPath:folderPath withIntermediateDirectories:YES attributes:nil error:nil];
    if ([fm fileExistsAtPath:path]) {
        [fm removeItemAtPath:path error:nil];
    }
}

@implementation LibraryImageSize

+ (RCTImageSource *)imageSourceWithPath:(NSString *)path {
    return [[RCTImageSource alloc] initWithURLRequest:[RCTConvert NSURLRequest:path]
                                                 size:CGSizeZero
                                                scale:1.0];
}

+ (nullable UIImage *)imageWithPath:(NSString *)path {
    RCTImageSource *source = [self imageSourceWithPath:path];
    if (!source || !source.request.URL || !source.request.URL.scheme) {
        return nil;
    }

    NSString *scheme = [source.request.URL.scheme lowercaseString];
    UIImage *image = nil;

    if ([scheme isEqualToString:@"file"]) {
        image = RCTImageFromLocalAssetURL(source.request.URL);
        if (!image) {
            image = RCTImageFromLocalBundleAssetURL(source.request.URL);
        }
    } else if ([scheme isEqualToString:@"data"] || [scheme hasPrefix:@"http"]) {
        NSData *data = [NSData dataWithContentsOfURL:source.request.URL];
        if (!data) return nil;
        return [UIImage imageWithData:data];
    }

    CGFloat scale = source.scale;
    if (scale == 1.0 && source.size.width > 0 && image) {
        scale = (CGFloat)CGImageGetWidth(image.CGImage) / source.size.width;
    }

    if (scale > 1.0 && image && image.CGImage) {
        image = [UIImage imageWithCGImage:image.CGImage
                                    scale:scale
                              orientation:image.imageOrientation];
    }

    return image;
}

+ (void)getSizesWithPaths:(NSArray<NSString *> *)paths
               completion:(void (^)(NSString *result))completion {
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
        NSMutableArray *sizes = [NSMutableArray array];

        for (NSString *path in paths) {
            if ([path hasPrefix:@"ph://"]) {
                PHAsset *asset = [MediaAssetManager fetchRawAssetWithIdentifier:path];
                if (asset) {
                    [sizes addObject:@{
                        @"width": @((double)asset.pixelWidth),
                        @"height": @((double)asset.pixelHeight)
                    }];
                }
                continue;
            }

            UIImage *image = [self imageWithPath:path];
            if (!image) continue;

            [sizes addObject:@{
                @"width": @(image.size.width),
                @"height": @(image.size.height)
            }];
        }

        NSError *error;
        NSData *jsonData = [NSJSONSerialization dataWithJSONObject:sizes options:0 error:&error];
        NSString *result = error ? @"[]" : [[NSString alloc] initWithData:jsonData encoding:NSUTF8StringEncoding];
        completion(result ?: @"[]");
    });
}

+ (nullable NSString *)saveImage:(UIImage *)image
                          format:(NSString *)format
                            path:(NSString *)path {
    ensurePath(path);

    NSURL *finalPath = toFilePath(path);
    NSData *data;

    if ([format isEqualToString:@"png"]) {
        data = UIImagePNGRepresentation(image);
        if (!data) return @"LibraryImageSize.failConvertToPNG";
    } else {
        data = UIImageJPEGRepresentation(image, 1.0);
        if (!data) return @"LibraryImageSize.failConvertToJPG";
    }

    NSError *error;
    BOOL success = [data writeToURL:finalPath options:NSDataWritingAtomic error:&error];
    if (!success || error) {
        return [NSString stringWithFormat:@"LibraryImageSize.catch.error: %@", error.localizedDescription];
    }

    return nil;
}

@end

File: ios/LibraryImageSize.swift
//
//  LibraryImageSize.swift
//  MediaLibrary
//
//  Created by sergeymild on 03/08/2023.
//  Copyright © 2023 Facebook. All rights reserved.
//

import Foundation
import React

private struct Size: Codable {
    let width: Double
    let height: Double
}

private func toFilePath(path: NSString) -> URL {
    if path.hasPrefix("file://") {
        return URL(string: path as String)!
    }
    return URL(string: "file://\(path)")!
}

@objc
public class LibraryImageSize: NSObject {
    private static func imageSource(path: String) -> RCTImageSource? {
        return RCTImageSource(
            urlRequest: RCTConvert.nsurlRequest(path),
            size: .zero,
            scale: 1.0
        )
    }
    
    @objc
    public static func image(path: String) -> UIImage? {
        guard let source = imageSource(path: path),
              let url = source.request.url,
              let scheme = url.scheme?.lowercased()
        else {
            return nil
        }
        
        var image: UIImage? = nil;
        if scheme == "file" {
            image = RCTImageFromLocalAssetURL(url) ?? RCTImageFromLocalBundleAssetURL(url)
        } else if scheme == "data" || scheme.hasPrefix("http") {
            guard let data = try? Data(contentsOf: url) else { return nil }
            return UIImage(data: data)
        }
        
        var scale = source.scale
        if scale == 1.0 && source.size.width > 0, let image = image {
          // If no scale provided, set scale to image width / source width
            scale = CGFloat(image.cgImage?.width ?? 1) / source.size.width;
        }
        
        if scale > 1.0, let i = image, let cgImage = i.cgImage {
            image = UIImage(
                cgImage: cgImage,
                scale: scale,
                orientation: i.imageOrientation
            )
        }
        
        return image
    }
    
    
    @objc
    public static func getSizes(paths: [String], completion: @escaping (String) -> Void) {
        Task {
            var sizes: [Size] = []
            for path in paths {
                if path.hasPrefix("ph://") {
                    if let asset = MediaAssetManager.fetchRawAsset(identifier: path) {
                        let size = Size(width: Double(asset.pixelWidth), height: Double(asset.pixelHeight))
                        sizes.append(size)
                    }
                    continue
                }
                
                guard let image = image(path: path) else {
                    continue
                }
                sizes.append(.init(width: image.size.width, height: image.size.height))
            }
            let result = String(data: (try! JSONEncoder().encode(sizes)), encoding: .utf8) ?? "[]"
            completion(result)
        }
    }
    
    @objc
    public static func save(image: UIImage, format: NSString, path: NSString) -> Bool {
        let folderPath = path.deletingLastPathComponent
        do {
            try FileManager.default.createDirectory(atPath: folderPath, withIntermediateDirectories: true)
            if FileManager.default.fileExists(atPath: path as String) {
                try FileManager.default.removeItem(atPath: path as String)
            }
            let finalPath = toFilePath(path: path)
            if format == "png" {
                guard let data = image.pngData() else { return false }
                try data.write(to: finalPath, options: .atomic)
                return true
            }
            guard let data = image.jpegData(compressionQuality: 1.0) else { return false }
            try data.write(to: finalPath, options: .atomic)
            return true
        } catch {
            debugPrint("save(image.Error", error)
        }
        return false
    }
}
